main/src/ca/uwaterloo/flix/language/phase/Uncurrier2.scala

```diff
@@ -350,17 +331,112 @@ object Uncurrier2 extends Phase[SimplifiedAst.Root, SimplifiedAst.Root] {
+        Term.Head.Lit(l, tpe, loc)
+      case Term.Head.App(exp, args, tpe, loc) =>
+        val e = uncurryCall(exp)
+        Term.Head.App(e, args, tpe, loc)
+    }
+
+    private def visitBodyTerm(term0: Term.Body): Term.Body = term0 match {
+      case Term.Body.Wild(tpe, loc) => term0
+      case Term.Body.QuantVar(sym, tpe, loc) => term0
+      case Term.Body.CapturedVar(sym, tpe, loc) => term0
+      case Term.Body.Lit(exp, tpe, loc) =>
+        val e = uncurryCall(exp)
+        Term.Body.Lit(e, tpe, loc)
+    }
+  }
+
+  /**
+   * Transform `defn0` into a series of definitions.
+   * Each definition in the series uncurries the previous definition one step further.
+   * The process stops when the body of the definition is no longer a lambda.
+   * @param ctxt0 A set of definitions and a mapping from a definition to its successor in the above series.
+   * @return `ctxt0` updated with the result of uncurrying `defn0`.
+   */
+  private def uncurryDefn(defn0: Def, ctxt0: UncurryContext)(implicit flix: Flix): UncurryContext =
+    // Flix assumes benchmarks and tests have a certain signature, so we leave those alone.
+    if (defn0.ann.isBenchmark || defn0.ann.isTest) {
+      ctxt0
+    } else {
+      defn0.exp match {
+        case Expression.Lambda(fparams, body, tpe, loc) =>
+          val sym1 = Symbol.freshDefnSym(defn0.sym)
+          val params0 = defn0.fparams ++ fparams
+          val params1 = params0.map(fparam => fparam.copy(sym = Symbol.freshVarSym(fparam.sym)))
+          val vars0 = params0.zip(params1).map {
+            case (oldParam, newParam) => oldParam.sym -> newParam.sym
+          }
+          val exp1 = SimplifiedAstOps.renameBoundVars(body, vars0.toMap)
+          val tpe1 = uncurryArrow(defn0.tpe)
+          val result = defn0.copy(sym = sym1, fparams = params1, exp = exp1, tpe = tpe1)
+          val defs1 = ctxt0.defs + (sym1 -> result)
+          val names1 = ctxt0.names + (defn0.sym -> sym1)
+          val ctxt1 = UncurryContext(defs1, names1)
+          uncurryDefn(result, ctxt1)
+
+        case _ => ctxt0
       }
     }
 
-    val result = root.copy(defs = uncurriedDefs1)
+  /** Uncurry an arrow type A -> (B -> C) into an arrow (A, B) -> C
+   * where A may be multiple types and A, B, C are all proper types.
+   * Example: (Int32 -> Option[Int32]) -> Int32 -> Option[Int32]
+   * uncurries into: (Int32 -> Option[Int32], Int32) -> Option[Int32]
+   */
+  private def uncurryArrow(tpe0: Type): Type = tpe0 match {
+    case Type.Apply(lhs0, Type.Apply(Type.Apply(Type.Arrow(2, eff1), b), c)) => //TODO: arity could be more than 2
+      def visitLhs(lhs0: Type): Type = lhs0 match {
+        case Type.Arrow(arity0, eff0) =>
+          val eff2 = eff1 match {
+            case Type.Impure => Type.Impure
+            case Type.Pure => eff0
+            case Type.Unit => Type.Unit //TODO: What does this mean?
+            case _ => throw InternalCompilerException(s"Unexpected effect: '${eff1.getClass}'.")
```

Comment:
After synchronizing with master, there is a test failure on IDE.flix where `eff1` is a type variable. The definition that triggers it is on line 69 in IDE.flix

Reply:
The solution was to use `mkAnd` from BoolUnification to combine the two effects. My match was just an incomplete version of that anyway.